src/scikit_build_core/resources/_editable_redirect.py

```diff
@@ -36,24 +36,68 @@ def __init__(
         self.verbose = verbose
         self.build_options = build_options
         self.install_options = install_options
+        # Construct the __path__ of all resource files
+        # I.e. the paths of all package-like objects
+        submodule_search_locations: dict[str, set[str]] = {}
+        pkgs: list[str] = []
+        # Loop over both python native source files and cmake installed ones
+        for tree in (known_source_files, known_wheel_files):
+            for module, file in tree.items():
+                # Strip the last element of the module
+                parent = ".".join(module.split(".")[:-1])
+                # Check if it is a package
+                if "__init__.py" in file:
+                    parent = module
+                    pkgs.append(parent)
+                # Skip if it's a root module (there are no search paths for these)
+                if not parent:
+                    continue
+                # Initialize the tree element if needed
+                submodule_search_locations.setdefault(parent, set())
+                # Add the parent path to the dictionary values
+                parent_path, _ = os.path.split(file)
+                if not parent_path:
+                    # root modules are skipped so all files should be in a parent package
+                    msg = f"Unexpected path to source file: {file} [{module}]"
+                    raise ImportError(msg)
+                if not os.path.isabs(parent_path):
+                    parent_path = os.path.join(str(DIR), parent_path)
+                submodule_search_locations[parent].add(parent_path)
+        self.submodule_search_locations = submodule_search_locations
+        self.pkgs = pkgs
 
     def find_spec(
         self,
         fullname: str,
         path: object = None,
         target: object = None,
     ) -> importlib.machinery.ModuleSpec | None:
+        # If current item is a know package use its search locations, otherwise if it's a module use the parent's
+        parent = (
+            fullname if fullname in self.pkgs else ".".join(fullname.split(".")[:-1])
+        )
+        # If no known submodule_search_locations is found, it means it is a root module. Do not populate its kwargs
+        # in that case
+        kwargs = {}
+        if parent in self.submodule_search_locations:
+            kwargs["submodule_search_locations"] = list(
+                self.submodule_search_locations[parent]
+            )
         if fullname in self.known_wheel_files:
             redir = self.known_wheel_files[fullname]
             if self.rebuild_flag:
                 self.rebuild()
+            # Note: MyPy reports wrong type for `submodule_search_locations`
```

Comment:
Where do we report this issue? Here is the snippet of its implementation:
```python
def spec_from_file_location(name, location=None, *, loader=None,
                            submodule_search_locations=_POPULATE):
    """Return a module spec based on a file location.

    To indicate that the module is a package, set
    submodule_search_locations to a list of directory paths.  An
    empty list is sufficient, though its not otherwise useful to the
    import system.

    The loader must take a spec as its only __init__() arg.

    """
    ...
    # Set submodule_search_paths appropriately.
    if submodule_search_locations is _POPULATE:
        # Check the loader.
        if hasattr(loader, 'is_package'):
            try:
                is_package = loader.is_package(name)
            except ImportError:
                pass
            else:
                if is_package:
                    spec.submodule_search_locations = []
    else:
        spec.submodule_search_locations = submodule_search_locations
    if spec.submodule_search_locations == []:
        if location:
            dirname = _path_split(location)[0]
            spec.submodule_search_locations.append(dirname)
```
The error reported is:
```
src/scikit_build_core/resources/_editable_redirect.py:98: error: Argument 3 to "spec_from_file_location" has incompatible type "**Dict[str, List[str]]"; expected "Optional[Loader]"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

Reply:
It's correct here too:

https://github.com/python/typeshed/blob/398b31de92fb84e9d594c257bd4efde46f2ee78a/stdlib/importlib/util.pyi#L33

So I think that's a mypy error. MyPy tends to hate `**`. If you make it a typed dict, it will be able to figure out that you got the types correct.

FYI, I tried our imports at the top, and modules like typing and pathlib are pulled in by `importlib.abc` (it's really heavy, actually, at 36 ms, more than Python's own startup time (31 ms if disable site.py, otherwise 46 ms). I think this is a pretty recent change, as 3.10 takes half the time (and doesn't pull in pathlib, though it does still pull in typing). Older versions actually take longer, though, because they import the smaller number of package much slower.

Reply:
Interesting. `importlib.abc` could be dropped or stripped out at installation. Don't suppose there's a tool for that?